Mathlib/NumberTheory/BernoulliPolynomials.lean

```diff
@@ -172,29 +208,77 @@ theorem bernoulli_succ_eval (n p : ℕ) : (bernoulli p.succ).eval (n : ℚ) =
   apply eq_add_of_sub_eq'
   rw [sum_range_pow_eq_bernoulli_sub]
 
-theorem bernoulli_eval_one_add (n : ℕ) (x : ℚ) :
-    (bernoulli n).eval (1 + x) = (bernoulli n).eval x + n * x ^ (n - 1) := by
+theorem bernoulli_comp_one_add_X (n : ℕ) :
+    (bernoulli n).comp (1 + X) = bernoulli n + n • X ^ (n - 1) := by
   refine Nat.strong_induction_on n fun d hd => ?_
-  have nz : ((d.succ : ℕ) : ℚ) ≠ 0 := by norm_cast
-  apply (mul_right_inj' nz).1
-  rw [← smul_eq_mul, ← eval_smul, bernoulli_eq_sub_sum, mul_add, ← smul_eq_mul, ← eval_smul,
-    bernoulli_eq_sub_sum, eval_sub, eval_finset_sum]
+  cases d with
+  | zero => simp
+  | succ d =>
+  have nz : (d + 2 : ℕ) ≠ 0 := by norm_cast
+  rw [← smul_right_inj nz, ← smul_comp, smul_add, bernoulli_eq_sub_sum', sub_comp, sum_comp]
+  simp only [add_assoc, one_add_one_eq_two, smul_smul]
   conv_lhs =>
-    congr
-    · skip
-    · apply_congr
-      · skip
-      · rw [eval_smul, hd _ (mem_range.1 (by assumption))]
-  rw [eval_sub, eval_finset_sum]
-  simp_rw [eval_smul, smul_add]
-  rw [sum_add_distrib, sub_add, sub_eq_sub_iff_sub_eq_sub, _root_.add_sub_sub_cancel]
-  conv_rhs =>
-    congr
+    enter [2]
+    apply_congr
     · skip
-    · congr
-      rw [succ_eq_add_one, ← choose_succ_self_right d]
-  rw [Nat.cast_succ, ← smul_eq_mul, ← sum_range_succ _ d, eval_monomial_one_add_sub]
-  simp_rw [smul_eq_mul]
+    · rw [smul_comp, hd _ (mem_range.1 (by assumption))]
+  simp_rw [smul_add, sum_add_distrib, sub_add, sub_add_eq_sub_sub_swap, sub_sub_eq_add_sub]
+  congr 1
+  rw [show ∀ a b c d : ℚ[X], a - b = c + d ↔ a - c = b + d by grind]
+  rw [smul_comp, ← smul_sub, X_pow_comp, one_add_X_pow_sub_X_pow]
+  trans ∑ i ∈ range (d + 1), ((d + 2).choose (i + 1) * (i + 1)) • X ^ i
+  · simp_rw [smul_sum, smul_smul, ← add_one_mul_choose_eq (d + 1), add_assoc, one_add_one_eq_two]
+  trans ∑ i ∈ range (d + 1),
+    ((d + 2).choose i * i) • X ^ (i - 1) + (((d + 2).choose (d + 1)) * (d + 1)) • X ^ (d + 1 - 1)
+  · rw [← sum_range_succ _ (d + 1)]
+    simp [sum_range_succ']
+  · simp [choose_succ_self_right, add_assoc, mul_assoc]
```

Comment:
This could perhaps be made more readable (although possibly a bit longer) by using `calc`, rather than these `trans` steps.